### PR TITLE
Added `PrepareEndChat` Telemetry Event to identify end chat workflow

### PR DESCRIPTION
**Pull request by charliewang95**
### Id of the task, bug, story or other reference.
[Task 3520706](https://dynamicscrm.visualstudio.com/OneCRM/_workitems/edit/3520706): CloseChat telemetry event unreliable
[Task 3568226](https://dynamicscrm.visualstudio.com/OneCRM/_workitems/edit/3568226): [V2] [Telemetry] Improve EndChatSDKCall trigger point logging

### Description
_Include a description of the problem to be solved_
While investigating telemetry, we see "EndChatSDKCall" event popping up without further details. This is triggered inside the endChat method, but there are many scenarios where this can happen, which caused investigation to be hard.

## Solution Proposed
_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_
Added `PrepareEndChat` Telemetry Event to identify end chat workflow, before each call of `endChat`. 

### Acceptance criteria
_Define what are the conditions to consider the PR has achieved the intended goal_
See screenshot of telemetry event and distinct description below

## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_
1. Customer ended chat with postchat configured
image

2. Bot ended chat with embed postchat configured, then customer closes the chat
image

3. Agent ended chat with embed postchat configured, then customer closes the chat
image

4. Agent ended chat with link postchat configured, then customer closes the chat
image

5. Customer ended chat without post chat configured
image

6. Session init succeeded but create adapter failed, then customer closes the chat
image

7. Received InitiateEndChat Broadcast event when link mode post chat is up (Inactive State)
image

8. During Multi-tab, tab1 ends chat in any way, tab 2 closes the chat as well
image

### Sanity Tests
-   [ ] You have tested all changes in Popout mode
-   [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
-   [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y
- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

__*Please provide justification if any of the validations has been skipped.*__

### CODE REVIEW

File: chat-widget/src/components/livechatwidget/common/endChat.ts [hunk 1]
```diff
@@ -29,8 +29,13 @@ const prepareEndChat = async (props: ILiveChatWidgetProps, chatSDK: any, state:
         if (conversationDetails?.canRenderPostChat?.toLowerCase() === Constants.false) {
             // If ended by customer, just close chat
             if (state?.appStates?.conversationEndedBy === ConversationEndEntity.Customer) {
+                TelemetryHelper.logSDKEvent(LogLevel.INFO, {
+                    Event: TelemetryEvent.PrepareEndChat,
+                    Description: "Conversation ended by customer. Post chat not configured or should not show."
```
Comment (xTEddie) [inline chat-widget/src/components/livechatwidget/common/endChat.ts:34]: Should these `Description` be refactored an external file for `ConversationEndReasonDescription`?

File: chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx [hunk 2]
```diff
@@ -404,6 +413,10 @@ export const LiveChatWidgetStateful = (props: ILiveChatWidgetProps) => {
 
         BroadcastService.getMessageByEventName(endChatEventName).subscribe((msg: ICustomEvent) => {
             if (msg?.payload?.runtimeId !== TelemetryManager.InternalTelemetryData.lcwRuntimeId) {
+                TelemetryHelper.logSDKEvent(LogLevel.INFO, {
+                    Event: TelemetryEvent.PrepareEndChat,
+                    Description: "Received EndChat BoradcastEvent from other tabs. Closing this chat."
```
Comment (xTEddie) [inline chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx:418]: nit: typo on `BoradcastEvent`

**Review by xTEddie — APPROVED** [review]

File: chat-widget/src/common/Constants.ts [hunk 3]
```diff
@@ -339,4 +339,17 @@ export class WidgetLoadTelemetryMessage {
...
+
+export class PrepareEndChatDescriptionConstants {
+    public static readonly ConversationEndedByCustomerWithoutPostChat = "Conversation ended by customer. Post chat not configured or should not show.";
+    public static readonly ConversationEndedByCustomerWithInvalidPostChat = "Conversation ended by customer. Post chat context is invalid.";
+    public static readonly ConversationEndedBy = "Conversation ended by";
+    public static readonly PrepareEndChatError = "There's an error while preparing to end chat. Closing chat widget.";
+    public static readonly WidgetLoadFailedAfterSessionInit = "SessionInit was successful, but widget load failed. Ending chat to avoid ghost chats in OC.";
+    public static readonly InitiateEndChatReceived = "Received InitiateEndChat BroadcastEvent while conversation state is not Active. Ending chat.";
+    public static readonly EndChatReceivedFromOtherTabs = "Received EndChat BroadcastEvent from other tabs. Closing this chat.";
+    public static readonly CustomerCloseChatOnFailureOrPostChat = "Cutsomer is trying to close chat widget on start chat failure or post chat pane.";
```
Comment (xTEddie) [inline chat-widget/src/common/Constants.ts:352]: nit: typo on `Cutsomer`

**Review by nareshme2016 — APPROVED** [review]

File: chat-widget/src/common/Constants.ts [hunk 4]
```diff
@@ -339,4 +339,17 @@ export class WidgetLoadTelemetryMessage {
...
+    public static readonly CustomerCloseInactiveChat = "Chat was Inactive and cutsomer is trying to close chat widget or refreshing the page.";
```
Comment (elopezanaya) [inline chat-widget/src/common/Constants.ts:353]: typo : cutsomer

File: docs/Telemetry.md [hunk 5]
```diff
@@ -273,6 +273,7 @@ Refer to the below table to understand different critical telemetry events raise
 |`CloseChatMethodException`|On failed to end chat from `ConfirmationPane` Confirm button click|
 |`GetChatReconnectContextSDKCallFailed`| On get reconnect context SDK call failed|
 |`GetConversationDetailsCallFailed`|On `PropactiveChatPane` load complete|
+|`PrepareEndChat`|On preparing to call ChatSDK end chat, used for identify end chat scenarios|
```
Comment (elopezanaya) [inline docs/Telemetry.md:276]: "used to identify end-chat scenarios"

**Review by elopezanaya — CHANGES_REQUESTED** [review]

**Review by elopezanaya — APPROVED** [review]
